Execute time-bound graceful shutdowns on `SIGINT`/`SIGTERM`.

Tener opened the pull request:
We introduce a delay of up to 3 seconds into the "immediate shutdown" path. If the graceful shutdown doesn't complete within this time, the process will exit anyway.

Changelog: Allocate 3 seconds for resource cleanup on "immediate" Teleport shutdown.

Fixes https://github.com/gravitational/teleport/issues/24968.

Tener committed:
Execute time-bound graceful shutdowns on SIGINT/SIGTERM.

We introduce a delay of up to 3 seconds into the "immediate shutdown"
path. If the graceful shutdown doesn't complete within this time, the process
will exit anyway.

Tener commented:
I realize this may be a controversial change, but I couldn't find a reliable way to do a shutdown cleanup otherwise. Adding a delay within individual services doesn't help: the process will simply exit, so the cleanups are interrupted in a hard way.

Tener committed:
Merge branch 'master' into tener/always-graceful-shutdowns

espadolini reviewed (commented):
There's _a lot_ of code that should be removed if we make this change, as nothing is supposed to be calling process.Close() anymore.

I'd also be wary of actually doing a graceful shutdown with a short time limit, since that codepath is basically never touched except during a SIGHUP upgrade, where the time limit is significantly more generous and the only practical thing that gets cancelled halfway through is `proxy.shutdown`.

Tener commented:
> There's _a lot_ of code that should be removed if we make this change, as nothing is supposed to be calling process.Close() anymore.

Agreed. However, I felt it was better to ask for opinions first, before gutting that part of the codebase.

> I'd also be wary of actually doing a graceful shutdown with a short time limit, since that codepath is basically never touched except during a SIGHUP upgrade, where the time limit is significantly more generous and the only practical thing that gets cancelled halfway through is `proxy.shutdown`.

I'm open to increasing this time limit, but I wouldn't make it too large. Maximum 10 seconds, I think? After that point, I doubt anything can be cleaned up in a time-bound manner.

There is a ton of things happening on shutdown, so I wasn't able to track everything, butit _feels_ safe: right now, SIGINT will perform almost no cleanup, and it gets called all the time. Allowing _some_ cleanup should be strictly better than _no_ cleanup.

smallinsky reviewed (commented):
https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/
>  Typically, with this graceful termination of the pod, kubelet makes requests to the container runtime to attempt to stop the containers in the pod by first sending a TERM (aka. SIGTERM) signal, with a grace period timeout, to the main process in each container. The requests to stop the containers are processed by the container runtime asynchronously. 

> When the grace period expires, the kubelet triggers forcible shutdown. The container runtime sends SIGKILL to any processes still running in any container in the Pod. The kubelet also cleans up a hidden pause container if that container runtime uses one.



Wonder what do you think about solution with separate  SIGTERM and SIGKILL flows: 

- SIGKILL -  will force "hard" process  exit by calling process.Close
- SIGTERM can trigger Graceful process shutdown with time limit

espadolini commented:
Our images already run teleport inside dumb-init, to replace the SIGTERM with a SIGQUIT to do graceful shutdown, and you can't catch SIGKILL, the kernel just stops all your threads and deletes your process anyway.

I like the idea of always doing at least a moderately graceful shutdown, I'm just worried about the consequences of effectively removing process.Close from the actually exercised code paths, and about the brand new introduction of a context that gets canceled in our graceful shutdowns - again, that's only ever done by canceling the shutdown after 5 minutes have passed during a SIGHUP restart, and not many people run those (and when they do, they often run into problems or hiccups).

Tener commented:
In a way, removing one of the ways we do shutdown would decrease the complexity. But I understand your concerns here; it's hard to say what is going to break in some edge cases... I feel we should go with it anyway (SIGKILL is always an option), but maybe there is something we can do to lower the risk of this change. I'm not sure what that would be, however.

smallinsky commented:
> Our images already run teleport inside dumb-init, to replace the SIGTERM with a SIGQUIT to do graceful shutdown, and you can't catch SIGKILL, the kernel just stops all your threads and deletes your process anyway.

@espadolini 
All right so the Kube deployment should already support auto db de-reregistration. But we want to also support SystemD deployment https://github.com/gravitational/teleport/blob/master/examples/systemd/teleport.service#L6
where we don't replace the `SIGTERM` 

> I like the idea of always doing at least a moderately graceful shutdown, I'm just worried about the consequences of effectively removing process.Close from the actually exercised code paths, and about the brand new introduction of a context that gets canceled in our graceful shutdowns 

We can always call  `process.Shutdown` and after shudownThreshold just call `process.Close()` and force the kill without relaying on context cancelation mechanism  but there might be some paths that doesn't fully respect `ctx.Done()`

Tener commented:
> We can always call `process.Shutdown` and after shudownThreshold just call `process.Close()` and force the kill without relaying on context cancelation mechanism but there might be some paths that doesn't fully respect `ctx.Done()`

I'd rather avoid calling `process.Close()` after `process.Shutdown()`. Already `.Shutdown()` is called in several places and deemed sufficient. It is hard to say if this would even be the correct usage of the API: I'd lean towards "no". 
 If we can eliminate all uses of `.Close()`, then I'd go for it.

> but there might be some paths that doesn't fully respect ctx.Done()

This isn't impossible, but probably unlikely. Whatever scheduling mechanism is involved it must take into account processes that will ignore signals whether due to bad behavior or simple bugs. For example, by default, Docker will wait 30s before going for a hard SIGKILL.

espadolini commented:
Can we do something similar to what we do during a SIGHUP, and give a graceful shutdown time of 3 seconds, then cancel the context, wait a couple more seconds for everything to finish, and then `os.Exit` if it hasn't finished yet? That would prevent infinite hangups due to bugs and wouldn't be worse than the current behavior in terms of graceful shutdowns.

Tener commented:
> Can we do something similar to what we do during a SIGHUP, and give a graceful shutdown time of 3 seconds, then cancel the context, wait a couple more seconds for everything to finish, and then `os.Exit` if it hasn't finished yet? That would prevent infinite hangups due to bugs and wouldn't be worse than the current behavior in terms of graceful shutdowns.

I think we should avoid `os.Exit()` given that this may not be what the caller of `WaitForSignals()` would have wanted, but I can add a fallback to handle possible bugs in timeout handling.

Tener committed:
Run `process.Shutdown()` in separate goroutine, to account for possible bugs.

Tener committed:
Support TELEPORT_UNSTABLE_SHUTDOWN_TIMEOUT env var.

Tener commented:
I've moved `process.Shutdown()` to a separate goroutine. Even if it blocks, the process will still exit after the timeout (https://github.com/gravitational/teleport/pull/31869/commits/4895943183d3c3df8f345e92e7d0461187a3cafc).

I've also added a new temporary, purposefully undocumented env var TELEPORT_UNSTABLE_SHUTDOWN_TIMEOUT, allowing override of the timeout period in the production setting. https://github.com/gravitational/teleport/pull/31869/commits/31d5475c205232fd3aa4cd0c17f8d1eb015757b4

Tener commented on lib/service/signals.go:
```diff
@@ -83,8 +83,33 @@ func (process *TeleportProcess) WaitForSignals(ctx context.Context) error {
+				cancelCtx, cancelFunc := context.WithTimeout(ctx, timeout)
+				process.log.Infof("Got signal %q, exiting within %vs.", signal, timeout.Seconds())
+				go func() {
+					process.Shutdown(cancelCtx)
```
Hmm, I'm not sure I'm following. We can't do anything with SIGKILL. What do you think should we do for `SIGTERM` and `SIGINT`?

smallinsky replied:
> We can't do anything with SIGKILL

AH, My bad. Than if this case handles only `syscall.SIGTERM` and `syscall.SIGINT` then this flow LGTM.

Tener replied:
Right, the switch looked like it was catching `SIGKILL`, but that wasn't truly the case. I have dropped `SIGKILL` from the case statement, it didn't do anything other than confuse people.

Tener committed:
Extract `getShutdownTimeout`, add tests.

Tener commented on lib/service/signals.go:
```diff
@@ -83,8 +83,33 @@ func (process *TeleportProcess) WaitForSignals(ctx context.Context) error {
+						}
+						timeout = t
+					}
+				}
```
Done.

espadolini commented on lib/service/signals.go:
```diff
@@ -83,8 +83,15 @@ func (process *TeleportProcess) WaitForSignals(ctx context.Context) error {
+				cancelCtx, cancelFunc := context.WithTimeout(ctx, timeout)
+				process.log.Infof("Got signal %q, exiting within %vs.", signal, timeout.Seconds())
+				go func() {
+					process.Shutdown(cancelCtx)
+					cancelFunc()
```
nit for slightly better panic safety
```suggestion
					defer cancelFunc()
					process.Shutdown(cancelCtx)
```

Tener committed:
Update lib/service/signals.go

Co-authored-by: Edoardo Spadolini <edoardo.spadolini@goteleport.com>

Tener committed:
Remove untrappable SIGKILL to reduce confusion.

Tener committed:
Merge branch 'master' into tener/always-graceful-shutdowns
